Replace guestfish with config drive on s390x platform

xkwangcn opened the pull request:
For remote mode, #174 is not supported.

This adds support for config-drive and replaces guestfish. 
#174 can only work well for local, but libguestfish is difficult and complicated to inject the Ignition config for remote libvirtd instance. 
This method uses  the OpenStack config-drive mechanism, which can support both local and remote instance.

xkwangcn commented:
The changes have already been verified using cmd/libvirt-actuator on s390x/mainframe platform.
and also been verified for openshift cluster, It could create the worker node successfully.:
`[root@test1-mff4j-bootstrap` ~]# oc get nodes
NAME                         STATUS     ROLES    AGE     VERSION
test1-mff4j-master-0         Ready      master   5h53m   v1.14.6+0ec926dcc
test1-mff4j-worker-0-mlrwq   NotReady   worker   41s     `v1.14.6+0ec926dcc`

xkwangcn commented:
@zeenix @praveenkumar, hi, could you have a look for this PR? 
If needed, I will modify my code in time when I get new comments from here. thank you~

zeenix commented:
> The changes have already been verified using cmd/libvirt-actuator on s390x/mainframe platform.
> and also been verified for openshift cluster, It could create the worker node successfully:

That's very great to hear. I'll review today.

P.S. please consider making use of ``` for pasting commandline and code. Saves many bytes on the internet. :)

xkwangcn commented:
> > The changes have already been verified using cmd/libvirt-actuator on s390x/mainframe platform.
> > and also been verified for openshift cluster, It could create the worker node successfully:
> 
> That's very great to hear. I'll review today.
> 
> P.S. please consider making use of ``` for pasting commandline and code. Saves many bytes on the internet. :)

ok, finished, it looks good~

xkwangcn commented on pkg/cloud/libvirt/client/s390_ignition.go:
```diff
@@ -1,25 +1,27 @@
 
 func setIgnitionForS390X(domainDef *libvirtxml.Domain, client *libvirtClient, ignition *providerconfigv1.Ignition, kubeClient kubernetes.Interface, machineNamespace, volumeName string) error {
-	glog.Info("Creating ignition file for s390x")
+	glog.Infof("Creating ignition file for s390x")
```
yes, unrelated change..., I will cancel this change.

zeenix replied:
please don't. The right thing to do would be to separate it in another patch.

xkwangcn replied:
Um, glog.Infof and glog.Info are both OK, right? I changed it because I want to unify to use glog.Infof in the file s390_ignition.go. 
Do I need to create a new PR to change this sentence? Or use glog.Infof directly here?

zeenix replied:
>  Do I need to create a new PR to change this sentence? Or use glog.Infof directly here?

No, I didn't say separate PR. :) I said, separate patch and yes, ideally every logical change (no matter how small) should go in a separate patch.

xkwangcn replied:
ok~, and could I submit a new commit in this PR? right?

xkwangcn replied:
I separated the patch by a new commit. any other advise pls let me know, thank you ~

xkwangcn commented on pkg/cloud/libvirt/client/domain.go:
```diff
@@ -36,7 +35,7 @@ type pendingMapping struct {
 }
 
-func newDomainDef() libvirtxml.Domain {
+func newDomainDef(arch string) libvirtxml.Domain {
```
This change's reason is that I think it may have an impact when targeting remote libvirtd instance, it related the xml which will be created for VM, and on the judgment of arch I think it should use getHostArchitecture not runtime. 
In fact, I don't know if "Graphics/Spice isn't supported on s390/s390x platform" is right, I see the code is first submitted in #174.
If you think it is not appropriate to put it here, I can put it in a new separate patch?

zeenix replied:
> For the judgment of arch, I think it should use getHostArchitecture not runtime.

I didn't disagree with that. I'm talking about passing around arch specifically instead of each call using `getHostArchitecture`. In future, these functions might need access to other data from connection and if we pass connection, we won't need to add more parameters.

zeenix replied:
>  If you think it is not appropriate to put it here, I can put it in a new separate patch?

Yes, ideally it should be a separate patch as its a separate logical change (will build and run fine without the config drive changes, can be depedency for other changes than this PR etc).

xkwangcn replied:
ok, I see, I agree with you. so I will create a new PR to modify this, and I will delete this change about newDomainDef here.

zeenix replied:
No separate PR needed for non-functional changes like this. Seperate patch/commit is warranted though.

xkwangcn replied:
sorry, I misunderstood what you meant. I can independently submit another commit in this PR, right?

xkwangcn replied:
I changed my code: 
pass "*libvirt.Connect" instead of arch returned by getHostArchitecture, and seperated the code as a new commit.
Is my understanding of your comments correct?

zeenix replied:
> I can independently submit another commit in this PR, right?

Yes. It's just `git push -f` as usual. :)

>  Is my understanding of your comments correct?

Yes. Thanks so much.

xkwangcn commented:
> Except for this minor improvement ideas and comment on commit log below, looks fine.
> 
> "cluster api to" part is redundant here since it's obvious which repo you're changing. `Use config drive ingnition for s390x` would be better, no? Also please mension in the details section that you're replacing `guestfish` usage.

ok, I have modified the  commit log, is it ok?

zeenix commented:
> ok, I have modified the commit log, is it ok?

Better but you still didn't add to details section.

xkwangcn commented:
> > ok, I have modified the commit log, is it ok?
> 
> Better but you still didn't add to details section.

ok, sorry, now I have added the details section. could you please help to review it?

zeenix commented:
> ok, sorry, now I have added the details section. could you please help to review it?

Thanks. Future you will thank you. :)

zeenix commented:
/ok-to-test

zeenix commented on pkg/cloud/libvirt/client/client.go:
```diff
@@ -194,8 +193,12 @@ func (client *libvirtClient) CreateDomain(input CreateDomainInput) error {
+	}
 	// Both "s390" and "s390x" are linux kernel architectures for Linux on IBM z Systems, and they are for 31-bit and 64-bit respectively.
-	if runtime.GOARCH == "s390x" || runtime.GOARCH == "s390" {
+	if arch == "s390x" || arch == "s390" {
```
I think changes to this file were supposed to be in the first patch as well. You can fix this with `git reset HEAD^` (don't hard reset), `git add PATH_TO_FILE`, `git commit --fix COMMIT_ID_OF_FIRST_COMMIT`, `git commit -a -c SHA1_OF_THIS_COMMIT` (you want to note down the SHA1 before doing any of these) and then an interactive rebase on master.

xkwangcn replied:
This change should be in the patch: Use "getHostArchitecture" instead of "runtime.GOARCH" to get the arch
Right?
I will add this change to the patch about arch.

zeenix replied:
> This change should be in the patch: Use "getHostArchitecture" instead of "runtime.GOARCH" to get the arch
> Right?

right.

xkwangcn replied:
done.

xkwangcn commented:
Hi, Zeeshan @zeenix 
now, "ci/prow/actuator-pkg-staleness — Job failed.".
I see the log about it, from the related file ```./hack/verify-actuator-pkg.sh```, It seems there's a missing file ```Gopkg.toml```. I have no idea how to modify my PR's code to pass it...
In my PR, could you give me some advise about how to pass the test? thank you very much~

xkwangcn commented:
2failling...
At present, I haven't found the cause of the failure. I will change it as soon as possible to pass the test.

zeenix commented:
> 2failling...

`actuator-pkg-staleness` is optional and it's known to fail so don't mind that one. :) No idea what `images` test fails. Let me rerun it:

/test images

xkwangcn commented:
> > 2failling...
> 
> `actuator-pkg-staleness` is optional and it's known to fail so don't mind that one. :) No idea what `images` test fails. Let me rerun it:
> 
> /test images

@zeenix thank you. now, only `actuator-pkg-stalenessso` failling and it's opetional. so, could this PR be merged now? Is there any problem I need to solve?

zeenix commented:
/lgtm
/approve

crawford commented:
/cherry-pick release-4.3
/cherry-pick release-4.2
